File: SelectionViewController/Classes/TDSelectionViewController.m
//
//  TDSelectionViewController.m
//  mCommerce
//
//  Created by Josh Campion on 10/07/2014.
//  Copyright (c) 2014 The Distance. All rights reserved.
//

#import "TDSelectionViewController.h"

@implementation TDSelectionViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    if (self.selectedKeys == nil) {
        self.selectedKeys = [NSMutableSet set];
    }
    
    // reset the selection type to assign the allowsMultipleSelection
    self.selectionType = self.selectionType;
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (self.selectedKeys.count > 0) {
        // preselect any cells already specified by the presenter
        for (id<NSCopying> key in self.selectedKeys) {
            
            NSIndexPath *selectedPath = [self indexPathForKey:key];
            if (selectedPath) {
                [self.tableView selectRowAtIndexPath:selectedPath animated:NO scrollPosition:UITableViewScrollPositionNone];
            }
        }
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void)setSelectionType:(SelectionType)selectionType
{
    _selectionType = selectionType;
    
    self.tableView.allowsMultipleSelection = selectionType != SelectionTypeSingle;
}

-(void)setOptions:(NSDictionary *)options withDetails:(NSDictionary *)details orderedAs:(NSArray *)orderedOptions
{
    self.sortedOptionKeys = orderedOptions;
    self.options = options;
    self.optionDetails = details;
}

-(void)cancelSelectionViewController:(id)sender
{
    [self.delegate selectionViewControllerRequestsCancel:self];
}

#pragma mark - Helper methods

-(NSIndexPath *)indexPathForKey:(id) key
{
    for (NSInteger s = 0; s < self.sortedOptionKeys.count; s++) {
        NSArray *section = self.sortedOptionKeys[s];
        for (NSInteger r = 0; r < section.count; r++) {
            if ([section[r] isEqual:key]) {
                return [NSIndexPath indexPathForRow:r inSection:s];
            }
        }
    }
    
    return nil;
}

-(id)keyForIndexPath:(NSIndexPath *) indexPath
{
    return self.sortedOptionKeys[indexPath.section][indexPath.row];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return self.sortedOptionKeys.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return ((NSArray *)self.sortedOptionKeys[section]).count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *cellIdentifier = [self tableView:tableView cellIdentifierForRowAtIndexPath:indexPath];
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    id<NSCopying> thisKey = [self keyForIndexPath:indexPath];
    NSString *thisValue = self.options[thisKey];
    NSString *thisDetail = self.optionDetails[thisKey];
    
    if ([cell conformsToProtocol:@protocol(SelectionCell)]) {
        UITableViewCell<SelectionCell>* selectionCell = (UITableViewCell<SelectionCell>*) cell;
        
        selectionCell.titleLabel.text = thisValue;
        selectionCell.detailLabel.text = thisDetail;
    }
    
    if ([self.selectedKeys containsObject:thisKey]) {
        //[tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
        //[cell setSelected:YES];
    }
    
    return cell;
}

-(NSString *)tableView:(UITableView *)tableView cellIdentifierForRowAtIndexPath:(NSIndexPath *) indexPath
{
    id<NSCopying> thisKey = [self keyForIndexPath:indexPath];
    NSString *thisDetail = self.optionDetails[thisKey];
    
    return (thisDetail != nil) ? @"Detail" : @"Basic";
}

-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return self.sectionTitles[section];
}

#pragma mark - Table view delegate

// Ensures the checkmarks are correctly displays and updates the users choices
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    // clear this re-selection
    id  selectedKey = self.sortedOptionKeys[indexPath.section][indexPath.row];
    
    if ([self.selectedKeys containsObject:selectedKey] && !self.requiresSelection) {
        [tableView deselectRowAtIndexPath:indexPath animated:YES];
        // manually send the delegate message as it doesn't get called when deselecting programmatically
        [tableView.delegate tableView:tableView didDeselectRowAtIndexPath:indexPath];
        return;
    }
    
    // if self.requiresSelection => should not be able to deselect - only allow user to deselect this choice by selecting another
    
    // if self.selected keys doesn't contain this key, this is a new selection. Continue below to conditionally deselect other elements
    
    // Perform any deselection
    if (self.selectionType == SelectionTypeSingle) {
        // other deselection is taken care of in tableView:didDeselectRowAtIndexPath: which is called as tableView.allowsMultipleSelection == false
    }
    
    if (self.selectionType == SelectionTypeSingleSectioned) {
        
        // clear an old value in this section manually as tableView.allowsMultipleSelection == true to allow selection in multiple sections
        
        // use a copy of the selected keys as this will change with any deselections
        NSSet *previousKeys = [self.selectedKeys copy];
        
        for (id key in previousKeys) {
            
            NSIndexPath *selectedPath = [self indexPathForKey:key];
            
            if (selectedPath.section == indexPath.section) {
                // clear the current selection to be none
                [tableView deselectRowAtIndexPath:selectedPath animated:YES];
                // manually send the delegate message as it doesn't get called when deselecting programmatically
                [tableView.delegate tableView:tableView didDeselectRowAtIndexPath:selectedPath];
            }
        }
    }
    
    if (self.selectionType == SelectionTypeMultiple) {
        // any number of selections can be made
    }
    
    if (self.selectionType == SelectionTypeMultipleSectioned) {
        // any number of selections can be made
    }
    
    [self.selectedKeys addObject:[self keyForIndexPath:indexPath]];
}

-(void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [self.selectedKeys removeObject:[self keyForIndexPath:indexPath]];
}

@end

File: SelectionViewController/Classes/TDSelectionViewController.swift
//
//  TDSelectionViewController.swift
//  SelectionViewController
//
//  Created by Josh Campion on 13/04/2016.
//  Copyright © 2016 The Distance. All rights reserved.
//

import Foundation

public extension TDSelectionViewController {
    
    /**
     
     Calculates the selected rows of `tableView` sorted by section. `nil` objects are entered where there are no selections for a given section. This can be used to validate the user's selection.
     
     - returns: An array of `Int`s representing a section, with either a selected `indexPath` for that section or `nil` if there is none. There is a tuple for each section in `tableView`.
     
    */
    public func sectionedSelections() -> [(Int, NSIndexPath?)] {
        
        let sectioned = self.tableView.indexPathsForSelectedRows?.map { ($0.section, $0) } ?? []
        
        var sectionedInfo = [Int:NSIndexPath?]()
        
        for (s, ip) in sectioned {
            sectionedInfo[s] = ip
        }
        
        for s in 0..<(sortedOptionKeys?.count ?? 0) {
            sectionedInfo[s] = sectionedInfo[s] ?? nil as NSIndexPath?
        }
        
        return sectionedInfo.map({ $0 }).sort({ $0.0 < $1.0 })
    }
    
    /**
     - returns: Flag for whether there is a selected `NSIndexPath` for each section in `tableView`.
     - seealso: `sectionedSelections()`.
    */
    public func validSectionedSelection() -> Bool {
        return sectionedSelections().reduce(true) { $0 && ($1.1 != nil) }
    }
    
    /**
     
     The error title shown if the user's selection is invalid given `selectionType` and `requiresSelection`.
     - returns: `nil`
    */
    public func errorTitle() -> String? {
        return nil
    }
    
    /**
     
     The error message shown if the user's selection is invalid given `selectionType == .Single / .Multiple` and `requiresSelection`.
     
     - returns: "Please make a selection."
    */
    public func errorMessageForInvalidSelection() -> String {
        return "Please make a selection."
    }
    
    /**
 
     The error message shown if the user's selection is invalid given `selectionType == .SingleSectioned / .MultipleSectioned` and `requiresSelection`. Default value is
     
     - reurns: "Please make a selection for " followed by a list of sections without selections.
    */
    public func errorMessageForInvalidSectionedSelection() -> String {
        
        let missingSelections = sectionedSelections().filter { $0.1 == nil }
            .map { self.tableView(self.tableView, titleForHeaderInSection: $0.0) ?? "Section \($0.0)" }
       .joinWithSeparator(", ")
        
        return "Please make a selection for \(missingSelections)."
    }
    
    /**
     
     The error title shown if the user's selection is invalid given `selectionType` and `requiresSelection`. 
     
     - returns: "OK".
    */
    public func errorDismissButtonTitle() -> String {
        return "OK"
    }
    
    /**
 
     Checks whether the view can be dismissed based on `selectionType` and `requiresSelection`. If it can dismissal is requested from the delegate otherwise a `UIAlertController` is shown.
     
     - seealso: `validSectionedSelection()`
     - seealso: `errorTitle()`
     - seealso: `errorMessageForInvalidSelection()`
     - seealso: `errorMessageForInvalidSectionedSelection()`
     - seealso: `errorDismissButtonTitle()`
    */
    @IBAction public func dismissSelectionViewController(sender:AnyObject?) {
     
        if (self.requiresSelection) {
            
            let validSelection:Bool
            
            switch (self.selectionType) {
            case .Single, .Multiple:
                validSelection = self.selectedKeys.count > 0;
            case .SingleSectioned, .MultipleSectioned:
                validSelection = self.validSectionedSelection()
            }
            
            if (validSelection) {
                self.delegate?.selectionViewControllerRequestsDismissal(self)
            } else {
                
                let alertMessage:String
                
                switch self.selectionType {
                case .Single, .Multiple:
                    alertMessage = errorMessageForInvalidSelection()
                case .SingleSectioned, .MultipleSectioned:
                    alertMessage = errorMessageForInvalidSectionedSelection()
                }
                
                let alert = UIAlertController(title: errorTitle(),
                                              message: alertMessage,
                                              preferredStyle: .Alert)
                
                alert.addAction(UIAlertAction(title: errorDismissButtonTitle(),
                    style: .Cancel,
                    handler: nil))
                
                self.presentViewController(alert, animated: true, completion: nil)
            }
        } else {
            self.delegate?.selectionViewControllerRequestsDismissal(self)
        }
    }
}